### ISSUE TITLE
Empty DatetimeIndex gets converted on assignment with .loc

### ISSUE THREAD

**Issue by tomasssalles** [pandas-dev/pandas#31222]
#### Code Sample

```python
from pandas import DataFrame, Series, DatetimeIndex
from numpy import float64

df = DataFrame(index=DatetimeIndex([]), columns=["A", "B"]).astype({"A": float64})
df.loc[:, "A"] = Series(index=df.index, dtype=float64)
type(df.index)
```

#### Output

```python
pandas.core.indexes.range.RangeIndex
```

#### Problem description

The index of the dataframe gets converted from `DatetimeIndex` into `RangeIndex` even though the `Series` assigned to one of the columns has the exact same index as the dataframe.

We often want to check that the index is a `DatetimeIndex` when validating the data we receive as input. If this can change unexpectedly from a simple assignment operation when the `DataFrame` is empty, it becomes necessary to treat empty `DataFrame`s as special cases in our code.

This is at the very least unexpected behaviour:
- It doesn't happen if the index is not empty;
- It doesn't happen if we use `df["A"]` (instead of `.loc`);
- It doesn't happen if there's only one column;
- It doesn't happen if we omit `.astype({"A": float64})` (i.e. if both columns initially have dtype `object`).

#### Expected Output

```python
pandas.core.indexes.datetimes.DatetimeIndex
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.6.9.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.3.0-26-generic
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8
pandas           : 0.25.3
numpy            : 1.18.1
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 19.3.1
setuptools       : 42.0.2
Cython           : None
pytest           : 5.3.2
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.10.3
IPython          : 7.11.1
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : 3.1.2
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : 1.3.12
tables           : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None

</details>

**Comment by TomAugspurger** [comment]
Can you try this with master (pandas 1.0.0rc0)? I think this is fixed.

**Comment by tomasssalles** [comment]
Confirmed: Fixed in 1.0.0rc0.
Thanks for the quick response!
